beps/bep_0007.rst

```diff
@@ -23,32 +24,35 @@ be interpreted as described in IETF `RFC 2119`_.
 
 .. _`RFC 2119`: http://tools.ietf.org/html/rfc2119
 
-Announce Parameter
-==================
+Announcing
+==========
+
+For each local IP address the client:
+
+	i. intends to receive incoming peer connections over (listens on)
+	ii. intends to publish to the tracker
+	iii. can route packets to the tracker via
 
-The client MAY add an ``&ipv6=`` parameter to the HTTP GET request it sends
-to the tracker. The value is either an IPv6 endpoint (address and port) or
-just an IPv6 address. In the case where only an address is supplied, the IPv6
-port is assumed to be the same as specified by the ``&port=`` parameter.
+an announce to the tracker SHOULD be made. Each announce should use the
+corresponding local IP address as the source address of the connection.
 
-The tracker SHOULD perform a NAT check on the IPv6 endpoint.
+The source address is set on a socket using the ``bind()`` call, before connecting.
 
-In case the client contacts the tracker on an IPv6 interface, it may add
-an ``&ipv4=`` parameter with its IPv4 address or endpoint. The value MUST
-be either an IPv4 endpoint (address and port) or just an IPv4 address. If only
-an address is supplied, the port is assumed to be the same as the ``&port=``
-parameter.
+This applies to TCP connections, for HTTP and HTTPS trackers, as well as UDP
+announces, to UDP trackers.
 
-The endpoints are encoded as strings as defined by `RFC 2732`_.
+If the tracker does not resolve to a an address family compatible with a source
+address, the client SHOULD NOT announce that address to the tracker.
 
-.. _`RFC 2732`: http://tools.ietf.org/html/rfc2732
+Point (iii) suggests the loopback address, ``127.0.0.1`` not be used as the
+source address when announcing to a tracker on the internet. As that source
+address cannot traverse the network.
```

Comment:
I don't quite get the purpose of this paragraph. If you're doing experimenting you might have a tracker on localhost. The same applies to link-local addresses. Clients should simply probe whether they can reach the tracker from each source address, it doesn't cost them much anyway.

Reply:
The paragraph is not saying you should never announce with a loopback address as the source, only when you're announcing to a tracker that's also not on the loopback "network". But I don't think this is quite correct anyway, a source address isn't a thing that traverses networks. But I'm not sure how to describe this.

Reply:
Linux has some crazy stuff that allows you to re-route global unicast network prefixes to the `lo` interface so that applications can [freebind](https://linux.die.net/man/7/ip) to any address in that range.

I don't think any special language should revolve around `lo` or the loopback address. Clients should just try to announce from any socket that they also bind to. It might fail, but that failure is not particularly costly.